internal/oci/store/overlay/store_overlay.go

```diff
@@ -0,0 +1,426 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package overlay implements an overlay based container store.
+package overlay
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	ociv1 "github.com/google/go-containerregistry/pkg/v1"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane/crossplane/internal/oci/layer"
+	"github.com/crossplane/crossplane/internal/oci/spec"
+	"github.com/crossplane/crossplane/internal/oci/store"
+)
+
+// Error strings
+const (
+	errMkContainerStore  = "cannot make container store directory"
+	errMkLayerStore      = "cannot make layer store directory"
+	errReadConfigFile    = "cannot read image config file"
+	errGetLayers         = "cannot get image layers"
+	errResolveLayer      = "cannot resolve layer to suitable overlayfs lower directory"
+	errCreateRootFS      = "cannot create OCI rootfs"
+	errWriteRuntimeSpec  = "cannot write OCI runtime spec"
+	errGetDigest         = "cannot get digest"
+	errMkAlgoDir         = "cannot create store directory"
+	errFetchLayer        = "cannot fetch and decompress layer"
+	errMkWorkdir         = "cannot create temporary work directory"
+	errApplyLayer        = "cannot apply (extract) uncompressed tarball layer"
+	errMvWorkdir         = "cannot move temporary work directory"
+	errStatLayer         = "cannot determine whether layer exists in store"
+	errCleanupWorkdir    = "cannot cleanup temporary work directory"
+	errMkOverlayDirTmpfs = "cannot make overlay tmpfs dir"
+)
+
+// Common overlayfs directories.
+const (
+	overlayDirTmpfs  = "tmpfs"
+	overlayDirUpper  = "upper"
+	overlayDirWork   = "work"
+	overlayDirLower  = "lower"  // Only used when there are no parent layers.
+	overlayDirMerged = "merged" // Only used when generating diff layers.
+)
+
+// Supported returns true if the supplied cacheRoot supports the overlay
+// filesystem. Notably overlayfs was not supported in unprivileged user
+// namespaces until Linux kernel 5.11. It's also not possible to create an
+// overlayfs where the upper dir is itself on an overlayfs (i.e. is on a
+// container's root filesystem).
+// https://github.com/torvalds/linux/commit/459c7c565ac36ba09ffbf
+func Supported(cacheRoot string) bool {
+	// We use NewLayerWorkdir to test because it needs to create an upper dir on
+	// the same filesystem as the supplied cacheRoot in order to be able to move
+	// it into place as a cached layer. NewOverlayBundle creates an upper dir on
+	// a tmpfs, and is thus supported in some cases where NewLayerWorkdir isn't.
+	w, err := NewLayerWorkdir(cacheRoot, "supports-overlay-test", []string{})
+	if err != nil {
```

Comment:
Would be helpful to capture/propagate this error message somehow for further investigation since there are a couple of cases where `NewLayerWorkdir` returns an error.

Reply:
What would we do with the error?

The purpose of this function is to determine whether overlayfs is supported or not. If it doesn't appear to be supported we don't return an error, we just fall back to using the more widely supported `store/uncompressed` package. It seems to me that the best we could do is log that we're not using overlayfs and why before falling back.

Reply:
That makes sense. I'm thinking in a followup PR it may be useful to add a flag to control this. That way we could default to the current behaviour, but also have a `--use-overlay-cache` (or something) that would actually return an error if it wasn't supported. This way folks could try force the overlay cache to find out whether it was going to work.